paramiko/hostkeys.py

```diff
@@ -196,14 +197,34 @@ def _hostname_matches(self, hostname, entry):
         """
         for h in entry.hostnames:
             if (
-                h == hostname or
+                self.wildcard_matches(h, hostname) or
                 h.startswith('|1|') and
                 not hostname.startswith('|1|') and
                 constant_time_bytes_eq(self.hash_host(hostname, h), h)
             ):
                 return True
         return False
 
+    def wildcard_matches(self, wildcard, hostname):
+        """
+        Implements '*' and '?' wildcards and '!' negation as in sshd.
+        :param str wildcard: host name / pattern from known_hosts
+        :param str hostname: target host name
+        :return bool:
+        """
+        wildcard_negation = False
+        if hostname.startswith('|1|'):
+            return False
+        if wildcard.startswith("!"):
+            wildcard_negation = True
+
+        wildcard = wildcard.lstrip("!")
```

Comment:
It would be more efficient to put this in the above if body and just slice off the first char:

```python
if wildcard.startswith("!"):
    wildcard_negation = True
    wildcard = wildcard[1:]
```